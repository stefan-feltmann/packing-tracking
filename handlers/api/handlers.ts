import { Context, APIGatewayProxyEvent } from 'aws-lambda'
import { createHash } from 'crypto'
import { sign, TokenExpiredError, verify } from 'jsonwebtoken'
import parseBearerToken from 'parse-bearer-token'

export const handlers = async (event: APIGatewayProxyEvent, _context: Context): Promise<any> => {
  // console.log(event.httpMethod)
  // console.log(_context)
  const method = event.httpMethod
  if (event && event.path && event.path !== '/v1/auth') {
    try {
      let tokenInfo = validateAuthToken(event)
    } catch (error) {
      console.log(error)
      switch (true) {
        case error instanceof TokenExpiredError:
          //TODO: Figure out why this isn't working
          let status = 401
          let tokenError = error as TokenExpiredError
          let errMsg = tokenError.toString()
          let headers = {
            headers: {
              'Content-Type': 'text/plain',
            },
          }
          let errorMsg = makeResponse(status, headers, errMsg)
          return errorMsg
          break
        default:
          let defaultHeaders = {
            headers: {
              'Content-Type': 'text/plain',
            },
          }
          let err = error as Error

          let errorDefaultMsg = makeResponse(500, defaultHeaders, err.message)
          return errorDefaultMsg
          break
      }
    }
  }
  switch (method) {
    case 'POST':
      switch (event.path) {
        case '/v1/auth':
          const authToken = getAuthToken(event)
          const outputAuth = outputStandard(authToken)
          return outputAuth
        case '/v1/user':
          const outputUser = outputDefault()
          return outputUser
        case '/v1/move':
          const outputMove = outputDefault()
          return outputMove

        default:
          const output = outputDefault()
          return output
      }
    default:
      const output = outputDefault()
      return output
  }
}

function makeResponse(status: number, headers: { headers: { 'Content-Type': string } }, errMsg: string) {
  return {
    statusCode: status,
    headers,
    body: errMsg,
  }
}

export function getAuthToken(event: any) {
  let token = ''
  try {
    const jwtSecret = getJwtSecret()
    const body = JSON.parse(event.body)

    //TODO: Use this for something
    const password = body.Password

    const returnData = { Username: body.Username }
    token = sign({ data: returnData }, jwtSecret, { expiresIn: '10m' })
  } catch (error) {
    console.error(error)
  }
  return token
}

export function validateAuthToken(event: any) {
  const jwtSecret = getJwtSecret()
  const rawToken = event.headers.Authorization
  const userToken = rawToken.replace('Bearer', '').trim()
  const token = verify(userToken, jwtSecret)
  return token
}

export function outputStandard(msgBody: string) {
  return {
    statusCode: 200,
    headers: {
      'Content-Type': 'text/plain',
    },
    body: msgBody,
  }
}

export function getJwtSecret() {
  // TODO: Setup a better secret
  const hash = createHash('sha512')

  hash.update('TODO: Setup a better secret')
  const hashDigest = hash.digest('hex')
  return hashDigest
}

function outputDefault() {
  return {
    statusCode: 200,
    headers: {
      'Content-Type': 'text/plain',
    },
    body: 'Hello, CDK!',
  }
}
