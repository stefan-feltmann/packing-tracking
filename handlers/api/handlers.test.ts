import { getAuthToken, outputStandard, handlers, validateAuthToken, getJwtSecret } from './handlers'
import { Context, APIGatewayProxyEvent } from 'aws-lambda'
import { verify, TokenExpiredError } from 'jsonwebtoken'
import { ContextProvider } from '@aws-cdk/core'

const LambdaTester = require('lambda-tester')

const password = getJwtSecret()

describe('API lambda Handlers', () => {
  test('getAuthToken', () => {
    let user = {
      Username: 'unit test',
      Password: 'unit password',
    }
    let event = { body: JSON.stringify(user) }
    let token = getAuthToken(event)
    const decodedToken = verify(token, password)
    expect(decodedToken).toMatchSnapshot({
      exp: expect.any(Number),
      iat: expect.any(Number),
    })
  })
  test('outputStandard', () => {
    let output = outputStandard('unit test')
    expect(output).toMatchSnapshot()
  })
  test('handlers', async () => {
    await LambdaTester(handlers)
      .event({ name: 'Unknown' })
      .expectResult((result: any) => {
        expect(result).toMatchSnapshot()
      })
    let user = {
      Username: 'unit test',
      Password: 'unit password',
    }
    await LambdaTester(handlers)
      .event({
        path: '/v1/auth',
        httpMethod: 'POST',
        headers: { User: 'test' },
        body: JSON.stringify(user),
      })
      .expectResult((result: any) => {
        expect(result).toMatchSnapshot({ body: expect.any(String) })
        let body = result.body
        const decodedToken = verify(body, password)
        expect(decodedToken).toMatchSnapshot({
          exp: expect.any(Number),
          iat: expect.any(Number),
        })
      })
    let user2 = {
      Username: 'unit test',
      Password: 'unit password',
    }
    let event = { body: JSON.stringify(user2) }
    let token = getAuthToken(event)
    await LambdaTester(handlers)
      .event({ path: '/v1/move', headers: { Authorization: `Bearer ${token}` } })
      .expectResult((result: any) => {
        expect(result).toMatchSnapshot()
      })
  })
  test('validateAuthToken', async () => {
    // let event = { headers: { User: 'unit test' } }
    let user2 = {
      Username: 'unit test',
      Password: 'unit password',
    }
    let event = { body: JSON.stringify(user2) }
    let token = getAuthToken(event)
    let testEvent = { headers: { Authorization: `Bearer ${token}` } }
    let result = validateAuthToken(testEvent)
    expect(result).toMatchSnapshot({
      exp: expect.any(Number),
      iat: expect.any(Number),
    })
    let oldToken =
      'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJkYXRhIjp7IlVzZXIiOiJ1bml0IHRlc3QifSwiaWF0IjoxNjMxMzMzMTUwLCJleHAiOjE2MzEzMzM3NTB9.mcqi6u8jHHf4iaXSKT40Ve6tM0oiNSwFMyqEQnM-MbE'
    // testEvent = { headers: { Authorization: `Bearer ${oldToken}` } }
    try {
      testEvent = { headers: { Authorization: `Bearer ${oldToken}` } }
      let result = validateAuthToken(testEvent)
      expect(result).toBeUndefined()
    } catch (error) {
      expect(error).toMatchSnapshot()
    }
  })
})
