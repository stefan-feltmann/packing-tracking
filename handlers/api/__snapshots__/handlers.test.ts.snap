// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`API lambda Handlers getAuthToken 1`] = `
Object {
  "data": Object {
    "Username": "unit test",
  },
  "exp": Any<Number>,
  "iat": Any<Number>,
}
`;

exports[`API lambda Handlers handlers 1`] = `
Object {
  "body": "Hello, CDK!",
  "headers": Object {
    "Content-Type": "text/plain",
  },
  "statusCode": 200,
}
`;

exports[`API lambda Handlers handlers 2`] = `
Object {
  "body": Any<String>,
  "headers": Object {
    "Content-Type": "text/plain",
  },
  "statusCode": 200,
}
`;

exports[`API lambda Handlers handlers 3`] = `
Object {
  "data": Object {
    "Username": "unit test",
  },
  "exp": Any<Number>,
  "iat": Any<Number>,
}
`;

exports[`API lambda Handlers handlers 4`] = `
Object {
  "body": "Hello, CDK!",
  "headers": Object {
    "Content-Type": "text/plain",
  },
  "statusCode": 200,
}
`;

exports[`API lambda Handlers outputStandard 1`] = `
Object {
  "body": "unit test",
  "headers": Object {
    "Content-Type": "text/plain",
  },
  "statusCode": 200,
}
`;

exports[`API lambda Handlers validateAuthToken 1`] = `
Object {
  "data": Object {
    "Username": "unit test",
  },
  "exp": Any<Number>,
  "iat": Any<Number>,
}
`;

exports[`API lambda Handlers validateAuthToken 2`] = `[TokenExpiredError: jwt expired]`;
